core/tokenrate/tokenrate.go

```diff
@@ -21,20 +22,46 @@ func init() {
 }
 
 func GetUSD(ctx context.Context, symbol string) (float64, error) {
-
+	mu := &sync.RWMutex{}
 	errs := make([]error, 0, len(quotes))
-	for _, q := range quotes {
-		r, err := q.getUSD(ctx, symbol)
-		if err == nil {
-			return r, nil
-		}
+	finalErrCh := make(chan error)
+	resultCh := make(chan float64)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+	}()
 
-		errs = append(errs, err)
+	for _, q := range quotes {
+		go func(q quoteQuery) {
+			errCh := make(chan error)
+			go q.getUSD(ctx, symbol, errCh, resultCh)
+			select {
+			case <-ctx.Done():
+				if ctx.Err() == context.DeadlineExceeded {
+					finalErrCh <- ctx.Err()
+				}
+				return
+			case e := <-errCh:
+				mu.Lock()
+				errs = append(errs, e)
+				mu.Unlock()
+				if len(errs) >= len(quotes) {
+					finalErrCh <- fmt.Errorf("%w: %s", ErrNoAvailableQuoteQuery, errs)
```

Comment:
only one message will be read from this channel, other routines with error will hang forever on write, since nobody will read msg. you need to create buffer for len(quotes) elements or create wait group to wait for all the tasks to finish

Reply:
Thanks for pointing out. I have made it simpler now. There was no need of second goroutine apparently, as it wasn't helping in anyway. Pls have another look and let me know.

Reply:
Also, there wasn't a need to listen for ctx.Done because of the way resty is designed. It returnes immediately after the parent context's deadline exceeds.